File: Provenance/Controller/PVNeoGeoPocketControllerViewController.m
//
//  PVNeoGeoPocketControllerViewController.m
//  Provenance
//
//  Created by Joe Mattiello on 03/06/17.
//  Copyright © 2017 James Addyman. All rights reserved.
//

#import "PVNeoGeoPocketControllerViewController.h"
#import <PVMednafen/MednafenGameCore.h>

@interface PVNeoGeoPocketControllerViewController ()

@end

@implementation PVNeoGeoPocketControllerViewController

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    for (JSButton *button in [self.buttonGroup subviews])
    {
        if (![button isMemberOfClass:[JSButton class]])
        {
            continue; // skip over the PVButtonGroupOverlayView
        }
        
        if ([[[button titleLabel] text] isEqualToString:@"A"])
        {
            [button setTag:OENGPButtonA];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"B"])
        {
            [button setTag:OENGPButtonB];
        }
    }
    
//    [self.leftShoulderButton setTag:PVPSXButtonL1];
//    [self.rightShoulderButton setTag:PVPSXButtonR1];

    [self.selectButton setTag:OENGPButtonOption];
    [self.startButton setTag:OENGPButtonOption];
}

- (void)dPad:(JSDPad *)dPad didPressDirection:(JSDPadDirection)direction
{
    MednafenGameCore *ngCore = (MednafenGameCore *)self.emulatorCore;
    
    [ngCore didReleaseNGPButton:OENGPButtonUp forPlayer:0];
    [ngCore didReleaseNGPButton:OENGPButtonDown forPlayer:0];
    [ngCore didReleaseNGPButton:OENGPButtonLeft forPlayer:0];
    [ngCore didReleaseNGPButton:OENGPButtonRight forPlayer:0];
    
    switch (direction)
    {
        case JSDPadDirectionUpLeft:
            [ngCore didPushNGPButton:OENGPButtonUp forPlayer:0];
            [ngCore didPushNGPButton:OENGPButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionUp:
            [ngCore didPushNGPButton:OENGPButtonUp forPlayer:0];
            break;
        case JSDPadDirectionUpRight:
            [ngCore didPushNGPButton:OENGPButtonUp forPlayer:0];
            [ngCore didPushNGPButton:OENGPButtonRight forPlayer:0];
            break;
        case JSDPadDirectionLeft:
            [ngCore didPushNGPButton:OENGPButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionRight:
            [ngCore didPushNGPButton:OENGPButtonRight forPlayer:0];
            break;
        case JSDPadDirectionDownLeft:
            [ngCore didPushNGPButton:OENGPButtonDown forPlayer:0];
            [ngCore didPushNGPButton:OENGPButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionDown:
            [ngCore didPushNGPButton:OENGPButtonDown forPlayer:0];
            break;
        case JSDPadDirectionDownRight:
            [ngCore didPushNGPButton:OENGPButtonDown forPlayer:0];
            [ngCore didPushNGPButton:OENGPButtonRight forPlayer:0];
            break;
        default:
            break;
    }
    
    [self vibrate];
}

- (void)dPadDidReleaseDirection:(JSDPad *)dPad
{
    MednafenGameCore *ngCore = (MednafenGameCore *)self.emulatorCore;
    
    [ngCore didReleaseNGPButton:OENGPButtonUp forPlayer:0];
    [ngCore didReleaseNGPButton:OENGPButtonDown forPlayer:0];
    [ngCore didReleaseNGPButton:OENGPButtonLeft forPlayer:0];
    [ngCore didReleaseNGPButton:OENGPButtonRight forPlayer:0];
}

- (void)buttonPressed:(JSButton *)button
{
    MednafenGameCore *ngCore = (MednafenGameCore *)self.emulatorCore;
    [ngCore didPushNGPButton:[button tag] forPlayer:0];
    
    [self vibrate];
}

- (void)buttonReleased:(JSButton *)button
{
    MednafenGameCore *ngCore = (MednafenGameCore *)self.emulatorCore;
    [ngCore didReleaseNGPButton:[button tag] forPlayer:0];
}

- (void)pressStartForPlayer:(NSUInteger)player
{
    MednafenGameCore *ngCore = (MednafenGameCore *)self.emulatorCore;
    [ngCore didPushNGPButton:OENGPButtonOption forPlayer:player];
}

- (void)releaseStartForPlayer:(NSUInteger)player
{
    MednafenGameCore *ngCore = (MednafenGameCore *)self.emulatorCore;
    [ngCore didReleaseNGPButton:OENGPButtonOption forPlayer:player];
}

- (void)pressSelectForPlayer:(NSUInteger)player
{
    MednafenGameCore *ngCore = (MednafenGameCore *)self.emulatorCore;
    [ngCore didPushNGPButton:OENGPButtonOption forPlayer:player];
}

- (void)releaseSelectForPlayer:(NSUInteger)player
{
    MednafenGameCore *ngCore = (MednafenGameCore *)self.emulatorCore;
    [ngCore didReleaseNGPButton:OENGPButtonOption forPlayer:player];
}

@end

File: Provenance/Controller/PVNeoGeoPocketControllerViewController.swift
//  Converted to Swift 4 by Swiftify v4.1.6640 - https://objectivec2swift.com/
//
//  PVNeoGeoPocketControllerViewController.swift
//  Provenance
//
//  Created by Joe Mattiello on 03/06/17.
//  Copyright © 2017 James Addyman. All rights reserved.
//

import PVMednafen

fileprivate extension JSButton {
    var buttonTag : PVNGPButton {
        get {
            return PVNGPButton(rawValue: tag)!
        }
        set {
            tag = newValue.rawValue
        }
    }
}

class PVNeoGeoPocketControllerViewController: PVControllerViewController<MednafenGameCore> {
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        buttonGroup?.subviews.forEach {
            guard let button = $0 as? JSButton, let text = button.titleLabel?.text else {
                return
            }
            if text == "A" {
                button.buttonTag = .A
            }
            else if text == "B" {
                button.buttonTag = .B
            }
        }

        selectButton?.buttonTag = .option
        startButton?.buttonTag = .option
    }

    override func dPad(_ dPad: JSDPad, didPress direction: JSDPadDirection) {
        let ngCore = emulatorCore
        ngCore?.didRelease(PVNGPButton.up, forPlayer: 0)
        ngCore?.didRelease(PVNGPButton.down, forPlayer: 0)
        ngCore?.didRelease(PVNGPButton.left, forPlayer: 0)
        ngCore?.didRelease(PVNGPButton.right, forPlayer: 0)
        switch direction {
            case .upLeft:
                ngCore?.didPush(PVNGPButton.up, forPlayer: 0)
                ngCore?.didPush(PVNGPButton.left, forPlayer: 0)
            case .up:
                ngCore?.didPush(PVNGPButton.up, forPlayer: 0)
            case .upRight:
                ngCore?.didPush(PVNGPButton.up, forPlayer: 0)
                ngCore?.didPush(PVNGPButton.right, forPlayer: 0)
            case .left:
                ngCore?.didPush(PVNGPButton.left, forPlayer: 0)
            case .right:
                ngCore?.didPush(PVNGPButton.right, forPlayer: 0)
            case .downLeft:
                ngCore?.didPush(PVNGPButton.down, forPlayer: 0)
                ngCore?.didPush(PVNGPButton.left, forPlayer: 0)
            case .down:
                ngCore?.didPush(PVNGPButton.down, forPlayer: 0)
            case .downRight:
                ngCore?.didPush(PVNGPButton.down, forPlayer: 0)
                ngCore?.didPush(PVNGPButton.right, forPlayer: 0)
            default:
                break
        }
        vibrate()
    }

    override func dPadDidReleaseDirection(_ dPad: JSDPad) {
        let ngCore = emulatorCore
        ngCore?.didRelease(PVNGPButton.up, forPlayer: 0)
        ngCore?.didRelease(PVNGPButton.down, forPlayer: 0)
        ngCore?.didRelease(PVNGPButton.left, forPlayer: 0)
        ngCore?.didRelease(PVNGPButton.right, forPlayer: 0)
    }

    override func buttonPressed(_ button: JSButton) {
        let ngCore = emulatorCore
        ngCore?.didPush(button.buttonTag, forPlayer: 0)
        vibrate()
    }

    override func buttonReleased(_ button: JSButton) {
        let ngCore = emulatorCore
        ngCore?.didRelease(button.buttonTag, forPlayer: 0)
    }

    override func pressStart(forPlayer player: Int) {
        let ngCore = emulatorCore
        ngCore?.didPush(PVNGPButton.option, forPlayer: UInt(player))
    }

    override func releaseStart(forPlayer player: Int) {
        let ngCore = emulatorCore
        ngCore?.didRelease(PVNGPButton.option, forPlayer: UInt(player))
    }

    override func pressSelect(forPlayer player: Int) {
        let ngCore = emulatorCore
        ngCore?.didPush(PVNGPButton.option, forPlayer: UInt(player))
    }

    override func releaseSelect(forPlayer player: Int) {
        let ngCore = emulatorCore
        ngCore?.didRelease(PVNGPButton.option, forPlayer: UInt(player))
    }
}
